Raise a exception when the number of intents is equal one

lucianaribeiro opened the pull request:
**Proposed changes**:
Closes [RasaHQ#4390](https://github.com/RasaHQ/rasa/issues/4390) 

- Raising exception when number of intents is less then 2 

**Changes made**:
- Added a Exception treating if is not possible to train in file `rasa/nlu/classifiers/embedding_intent_classifier.py` 

**Expected Behavior**:

- The number of intents must be less the 2 in your project file `/nlu`;
- Start training 
- The exception will raise  

**Status**:
- [ ] made PR ready for code review
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

arthurTemporim commented on rasa/nlu/classifiers/embedding_intent_classifier.py:
```diff
@@ -520,11 +520,15 @@ def train(
+            #     "Can not train a classifier. "
+            #     "Need at least 2 different classes. "
+            #     "Skipping training of classifier."
+            # )
```
If this piece of code is not relevant anymore should be removed.

arthurTemporim commented on rasa/nlu/classifiers/embedding_intent_classifier.py:
```diff
@@ -520,11 +520,15 @@ def train(
+            #     "Need at least 2 different classes. "
+            #     "Skipping training of classifier."
+            # )
+            raise Exception("Can not train a classifier. "
```
Maybe this is not the best solution. Because you can train your chatbot even without a single intent. Looking on other `raise` examples in this class, appears to be used only if the "problem" is big enough to disable the training. Is there anything that I am missing?

wochinge committed:
remove unused imports

joaorobson committed:
Add method to validate if domain is valid

joaorobson committed:
Test verify_domain method

joaorobson committed:
Exit early if domain is not alright

joaorobson committed:
Test early exiting of validate_files method

joaorobson committed:
Update CHANGELOG and documentation

wochinge committed:
use fallback tracker store in case of error during init

wochinge committed:
add FailSafeTrackerStore

wochinge committed:
wrap tracker store with FailSafeTrackerStore

wochinge committed:
fix typo in variable name

wochinge committed:
keep old tracker store

This means Rasa can come back to use the old tracker store again in case it comes up again.

wochinge committed:
update changelog

wochinge committed:
add test for error during __init__

wochinge committed:
add types

federicotdn committed:
Update Makefile

erohmensing committed:
update black version

erohmensing committed:
Update requirements-dev.txt

Co-Authored-By: ricwo <ric.wkr@gmail.com>

erohmensing committed:
Merge pull request #4702 from RasaHQ/update-black

update black version

jplpereira committed:
Added another logic value in one of the path ifs

Co-authored-by: Icaro Aragao <icaropsa@gmail.com>

jplpereira committed:
Add fix to changelog

federicotdn committed:
Merge pull request #4700 from RasaHQ/prepare-tests-macos

Install wget on macOS during tests setup

tmbo committed:
Merge pull request #4634 from RasaHQ/update_mattermost_docs

minor updates to the mattermost connector page

ricwo committed:
wip

ricwo committed:
f-strings

ricwo committed:
add test

Commit:
document AugmentedMemoizationPolicy

tmbo committed:
Merge branch 'master' into hsm207

imLew committed:
Merge branch 'master' into warnings-in-tests

imLew committed:
do not fail on warnings yet

ricwo committed:
fix test

ricwo committed:
add E231 exception

ricwo committed:
newline

ricwo committed:
Merge branch 'use-f-strings' into rabbit-app-id-property

Commit:
Update rasa/core/channels/slack.py

Co-Authored-By: Tobias Wochinger <t.wochinger@rasa.com>

jplpereira committed:
Merge branch 'master' into fix-4694

Commit:
Apply suggestions from code review

typos and f-strings messed up by black

Co-Authored-By: Tobias Wochinger <t.wochinger@rasa.com>

ricwo committed:
Merge pull request #4707 from RasaHQ/use-f-strings

f-strings

Commit:
Update rasa/nlu/training_data/training_data.py

Co-Authored-By: Tobias Wochinger <t.wochinger@rasa.com>

imLew committed:
logs and warnings review, + MDB tracker type hints

imLew committed:
Merge branch 'warnings-in-tests' of github.com:RasaHQ/rasa into warnings-in-tests

erohmensing commented:
Hey there @lucianaribeiro, not sure if you were trying to open a PR against your own repo or not -- in order to get it into our code and close the issue, you'd have to open the PR [here](https://github.com/RasaHQ/rasa/pulls). But perhaps it's best to wait until @arthurTemporim approves :)

lucianaribeiro commented:
Hi @erohmensing I have opened the PR against my own repo intentionally, only to @arthurTemporim take a look and help me with any problem that I may have. As soon as I finish it I will open the PR in Rasa's repo! Thanks for the support ;))

ricwo committed:
Merge branch '1.4.x' into rabbit-app-id-property

erohmensing committed:
Update CHANGELOG.rst

erohmensing committed:
Merge pull request #4696 from RasaHQ/export-nlu-as-json-from-interactive

Export nlu as json from interactive

ricwo committed:
Merge branch '1.4.x' into rabbit-app-id-property

ricwo committed:
merge 1.4.x

Commit:
Support Markdown sections where the title contains the delimiter (:)

ricwo committed:
Update rasa/core/brokers/pika.py

Co-Authored-By: Tobias Wochinger <t.wochinger@rasa.com>

jplpereira committed:
Update rasa/cli/scaffold.py

Co-Authored-By: Ella Rohm-Ensing <erohmensing@gmail.com>

ricwo committed:
Merge pull request #4715 from RasaHQ/rabbit-app-id-property

RabbitMQ app_id message property

wochinge committed:
Merge branch 'master' into granular-fingerprint

wochinge committed:
provide default value for `on_tracker_store_error`

This allows to make `Agent.create_tracker_store` static again.

wochinge committed:
Merge branch 'master' into tracker-fallback

eric-bonfadini committed:
Upgraded jsonschema version

joaorobson committed:
Merge domain validation methods

wochinge committed:
Merge pull request #4251 from RasaHQ/granular-fingerprint

Avoid retraining Core model if only templates section is changed

wochinge committed:
Merge branch 'master' into tracker-fallback

wochinge committed:
install requirements from requirements.txt for doc tests

erohmensing committed:
Merge pull request #4727 from RasaHQ/master-fix

fix failing docs on master

wochinge committed:
Merge branch 'master' into tracker-fallback

erohmensing committed:
update tensorflow version

Commit:
Merge branch 'master' into warnings-in-tests

wochinge committed:
Merge pull request #4699 from RasaHQ/tracker-fallback

Tracker fallback

wochinge committed:
Merge branch 'master' into master_issue-4512

erohmensing committed:
install requirements for docs stage

joaorobson committed:
Merge branch 'master' into exit_early_if_domain_empty

tschnoelzer committed:
buttons are optional in elements.

erohmensing committed:
Merge pull request #4732 from RasaHQ/reqs-docs

install requirements for docs stage

joaorobson committed:
Merge branch 'master' into exit_early_if_domain_empty

erohmensing committed:
update tensorflow requirement

eric-bonfadini committed:
Merge remote-tracking branch 'upstream/master' into master_issue-4512

eric-bonfadini committed:
Updated changelog

eric-bonfadini committed:
Merge branch 'master_issue-4512' of github.com:eric-bonfadini/rasa into master_issue-4512

erohmensing committed:
port travis doc fixes from master

erohmensing committed:
Merge pull request #4724 from eric-bonfadini/master_issue-4512

Upgraded jsonschema version

ricwo committed:
Merge branch 'master' into master

tschnoelzer committed:
added fixed comment to changelog

tschnoelzer committed:
Merge branch 'master' of https://github.com/tschnoelzer/rasa

erohmensing committed:
Merge pull request #4722 from RasaHQ/tensorflow-warning-pip

remove numpy deprecation errors when installed from pip

jplpereira committed:
Moved changelog entry

joaorobson committed:
Merge branch 'master' into exit_early_if_domain_empty

imLew committed:
Merge branch 'master' into warnings-in-tests

imLew committed:
adapt test to account for change in logs/warnings

imLew committed:
remove non-functioning test

ArjaanBuijk committed:
Add rasa x --config CONFIG (#4223)

Commit:
Update docs/core/policies.rst

Co-Authored-By: Vladimir Vlasov <mr.voov@gmail.com>

Commit:
simplify text in line with edit requests

jplpereira committed:
Moved changelog entry

Co-Authored-By: Ella Rohm-Ensing <erohmensing@gmail.com>

joaorobson committed:
Fix docstring and update documentation of verify_domain_validity

joaorobson committed:
Fix import and change to f-string

joaorobson committed:
Merge branch 'exit_early_if_domain_empty' of https://github.com/GCES-2-2019-Rasa/rasa into rasa_master

erohmensing committed:
Merge pull request #4661 from GCES-2-2019-Rasa/exit_early_if_domain_empty

Exit early if domain file is invalid

erohmensing committed:
Merge branch 'master' into fix-4694

akelad committed:
Merge branch 'master' into augmented-memo-docs

tabergma committed:
Merge pull request #4718 from yulkes/bugfix_allow_colon_in_section_title

Support Markdown sections where the title contains the delimiter (:)

erohmensing committed:
Merge pull request #4711 from GCES-2-2019-Rasa/fix-4694

remove error traceback on rasa init keyboard interrupt

akelad committed:
Merge branch 'master' into augmented-memo-docs

akelad committed:
Merge pull request #4708 from ortsaCniveK/augmented-memo-docs

document AugmentedMemoizationPolicy

Ghostvv committed:
use only word level featurizer for intents. add test

Ghostvv committed:
black

Ghostvv committed:
update changelog

Ghostvv committed:
remove exit

Ghostvv committed:
Merge pull request #4743 from RasaHQ/fix_intent_featurizer

use only word level featurizer for intents

wochinge committed:
Merge branch 'master' into 1.4.x-merge

erohmensing committed:
remove unreleased patches from changelog

erohmensing committed:
Merge pull request #4738 from RasaHQ/1.4.x-merge

merge 1.4.x into master

hsm207 committed:
Merge branch 'master' into hsm207

hsm207 committed:
Fix result returned by sanitize_examples()

Sanitize example should return a list of messages whose
order should follow its inputs. See https://github.com/RasaHQ/rasa/pull/4414#issuecomment-549334785 for details.

hsm207 committed:
Fix import error

hsm207 committed:
Fix to make sanitize_examples() be able to sort tokens

hsm207 committed:
Added unit test to test Token comparisons

hsm207 committed:
Fix typo

Co-Authored-By: Tanja <tabergma@gmail.com>

hsm207 committed:
Code style changes

Based on feedback from @federicotdn

erohmensing committed:
Merge branch 'master' into master

erohmensing committed:
Merge pull request #4735 from tschnoelzer/master

buttons are optional in elements.

hsm207 committed:
Updated Token comparison implementation

Based on feedback from @federicotdn

hsm207 committed:
Revert "Refactored fix for issue #1446"

This reverts commit 0330f264609b77e22ca5260e343ce564fb0b8ee9.

hsm207 committed:
Merge branch 'master' into hsm207

amn41 committed:
change headline of checkpoints section

lucianaribeiro committed:
Changing exception to ValueError

amn41 committed:
Merge pull request #4758 from RasaHQ/change-checkpoints-headline

change headline of checkpoints section

indam23 committed:
recognise e2e format when intent restated as /intent

lucianaribeiro committed:
Deleting comments

wochinge committed:
add types to the `MarkdownWriter` class

tmbo committed:
dump domain allowing unicode chars

wochinge committed:
quote `TrainingData` and `Message` because it's only imported for type checking

tmbo committed:
added a test

wochinge committed:
fix type (message is a dict)

wochinge committed:
Merge pull request #4761 from RasaHQ/training-data-types

add types to the `MarkdownWriter` class

erohmensing committed:
update pip before installing

erohmensing committed:
update base dockerfile

tmbo committed:
fixed comment screwing up file parsing

tmbo committed:
Merge pull request #4414 from hsm207/hsm207

Remove duplicate examples when creating TrainingData

tmbo committed:
Merge pull request #4762 from RasaHQ/unicode-domain-dumping

dump domain allowing unicode chars

tmbo committed:
prepared next version

tmbo committed:
merged 1.4.x

dakshvar22 committed:
added new featurizer

erohmensing committed:
only log unseen features if domain is not empty

erohmensing committed:
changelog

shubham3582 committed:
Bug fixed - #4710

shubham3582 committed:
Bug fixed - #4710

erohmensing committed:
handle when domain is none, refactor

erohmensing committed:
fix tests

ArjaanBuijk committed:
Update CHANGELOG.rst

Co-Authored-By: Tobias Wochinger <t.wochinger@rasa.com>

ArjaanBuijk committed:
Pass validated config_path

erohmensing committed:
Merge pull request #4766 from RasaHQ/parse-warning

only log unseen features if domain is not empty

dakshvar22 committed:
added batcing and fixed featurizer

indam23 committed:
Update rasa/nlu/training_data/loading.py

Co-Authored-By: Tobias Wochinger <mail@tobias-wochinger.de>

indam23 committed:
Update changelog for e2e format recognition fix

dakshvar22 committed:
formatter, removed prints

erohmensing committed:
changelog

erohmensing committed:
Merge branch '1.4.x' into dockerfiles

erohmensing committed:
Merge pull request #4763 from RasaHQ/dockerfiles

fix missing tensorflow version in dockerfiles

imLew committed:
Merge branch 'master' into warnings-in-tests

imLew committed:
fix lingering merge conflict

wochinge committed:
Merge pull request #4760 from melindaloubser1/e2e_format_recognition

recognise e2e format when intent restated as /intent

erohmensing committed:
prepare 1.4.5 release

Commit:
Update rasa/core/interpreter.py

Co-Authored-By: Tobias Wochinger <t.wochinger@rasa.com>

Commit:
Update rasa/core/domain.py

Co-Authored-By: Tobias Wochinger <t.wochinger@rasa.com>

Commit:
Update rasa/core/agent.py

Co-Authored-By: Tobias Wochinger <t.wochinger@rasa.com>

Commit:
Apply suggestions from code review

Co-Authored-By: Tobias Wochinger <t.wochinger@rasa.com>

imLew committed:
suggestions from code review

imLew committed:
Merge branch 'warnings-in-tests' of github.com:RasaHQ/rasa into warnings-in-tests

imLew committed:
formatting

wochinge committed:
Merge pull request #4771 from RasaHQ/release-1.4.x

prepare 1.4.5 release

erohmensing committed:
Merge branch 'master' into 1.4.x-merge

imLew committed:
revert tracker warning in test and fix validator tests

erohmensing committed:
Merge pull request #4773 from RasaHQ/1.4.x-merge

1.4.x merge

ArjaanBuijk committed:
Update CHANGELOG.rst

Co-Authored-By: Tobias Wochinger <t.wochinger@rasa.com>

ArjaanBuijk committed:
Update rasa/cli/x.py

Co-Authored-By: Tobias Wochinger <t.wochinger@rasa.com>

ArjaanBuijk committed:
Import module, not its functions

ArjaanBuijk committed:
Pass config_path as the last argument

This allows rasa x to remain compatible with older versions of rasa

ArjaanBuijk committed:
Merge remote-tracking branch 'origin/4223-r2' into 4223-r2

imLew committed:
change errors in validator to warnings

imLew committed:
Merge branch 'master' into warnings-in-tests

federicotdn committed:
Make sure event broker is correctly configured for Rasa X (local mode)

federicotdn committed:
Update changelog.

imLew committed:
merge unseen intent and entity test into feature

imLew committed:
change logger to warnings

imLew committed:
no longer set log level to debug for all tests

imLew committed:
test for warnings instead of logs

imLew committed:
restore log level to debug

federicotdn committed:
Fix changelog entry.

dakshvar22 committed:
added requirements

imLew committed:
formatting

imLew committed:
do not count debug messages

Commit:
Merge pull request #4673 from RasaHQ/warnings-in-tests

Change warnings to python standard

federicotdn committed:
Merge pull request #4779 from RasaHQ/event-broker-rasa-x

Fix Rasa X (local) not working when any tracker store was configured for Rasa

erohmensing committed:
update versioning info and name

erohmensing committed:
Merge pull request #4785 from RasaHQ/update-version

update python and pip versions in installation docs

shubham3582 committed:
Bug Fix #4710

shubham3582 committed:
Format rasa/core/processor.py

arthurTemporim reviewed (approved):
To me is an acceptable solution :D

wochinge committed:
Use the matplotlib backend ``agg`` in case the ``tkinter`` package is not installed

wochinge committed:
remove unused imports

dakshvar22 committed:
added new requirements

dakshvar22 committed:
Merge branch 'master' into convert_feat

wochinge committed:
Merge pull request #4791 from RasaHQ/tkinter-fix

Tkinter fix

dakshvar22 committed:
added tests, requirements and compatible with response selector

dakshvar22 committed:
fix tests

dakshvar22 committed:
fix deps

dakshvar22 committed:
fix deps

wochinge committed:
Merge branch 'master' into 4223-r2

dakshvar22 committed:
Update CHANGELOG.rst

Co-Authored-By: Vladimir Vlasov <vladimir@rasa.com>

dakshvar22 committed:
added docs, added tests, changed some constants

dakshvar22 committed:
merge conflict

dakshvar22 committed:
Update docs/nlu/choosing-a-pipeline.rst

Co-Authored-By: Vladimir Vlasov <vladimir@rasa.com>

dakshvar22 committed:
Update docs/nlu/components.rst

Co-Authored-By: Vladimir Vlasov <vladimir@rasa.com>

dakshvar22 committed:
updated docs, changed logic

dakshvar22 committed:
fix conflicts

dakshvar22 committed:
cosmetic changes

dakshvar22 committed:
cosmetic fixes

dakshvar22 committed:
update changelog

ArjaanBuijk committed:
Merge pull request #4741 from RasaHQ/4223-r2

Add rasa x --config CONFIG

dakshvar22 committed:
fix conflicts in changelog

federicotdn committed:
Allow metadata to be attached to any tracker event.

federicotdn committed:
Fix types and tests.

dakshvar22 committed:
Update CHANGELOG.rst

Co-Authored-By: Vladimir Vlasov <vladimir@rasa.com>

erohmensing committed:
Merge branch 'master' into master

dakshvar22 committed:
revert back constants

dakshvar22 committed:
fix constants

erohmensing committed:
fix indent and add test

erohmensing committed:
black

dakshvar22 committed:
removed versions from deps

erohmensing committed:
Merge pull request #4786 from shubham3582/master

stop default intents from raising undefined parsing error

federicotdn committed:
Fix three more tests.

federicotdn committed:
Black formatting.

dakshvar22 committed:
changed extra requires

dakshvar22 committed:
Merge branch 'master' of github.com:RasaHQ/rasa into convert_feat

federicotdn committed:
Remove unused copy of parameters.

dakshvar22 committed:
added note about optional requirements in docs

federicotdn committed:
Update changelog.

federicotdn committed:
Merge branch 'master' into events-metadata

federicotdn committed:
Add types.

federicotdn committed:
Modify comment about pickled trackers.

federicotdn committed:
Return metadata in Event.as_dict() only if metadata is present.

(Except for UserUttered and BotUttered)

federicotdn committed:
Black formatting.

dakshvar22 committed:
force push to restart

federicotdn committed:
Fix test.

dakshvar22 committed:
Merge pull request #4800 from RasaHQ/convert_feat

Add new featurizer based on ConveRT model

federicotdn committed:
Merge branch 'master' into events-metadata

federicotdn committed:
Merge pull request #4813 from RasaHQ/events-metadata

Allow metadata to be attached to any tracker Event

federicotdn committed:
Prepare 1.5.0 release.

wochinge committed:
dump NLU data in original order

wochinge committed:
Revert "dump NLU data in original order"

This reverts commit 53c7b029378cb89e9b76a01b055bd1a1993a9d83.

federicotdn committed:
Prepare 1.4.6 patch release.

tmbo committed:
Merge pull request #4824 from RasaHQ/release-1.4.6

Prepare 1.4.6 patch release

federicotdn committed:
Merge branch '1.4.x' into prepare-1.5.0

federicotdn committed:
Update Sanic version.

federicotdn committed:
Fix Sanic version update.

federicotdn committed:
Merge pull request #4822 from RasaHQ/prepare-1.5.0

Prepare 1.5.0 release

lucianaribeiro committed:
Merge pull request #8 from RasaHQ/master

Updating branch
